src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -42,51 +42,124 @@ ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
   _heap(heap),
   _mutator_free_bitmap(max_regions, mtGC),
   _collector_free_bitmap(max_regions, mtGC),
+  _old_collector_free_bitmap(max_regions, mtGC),
   _max(max_regions)
 {
   clear_internal();
 }
 
-void ShenandoahFreeSet::increase_used(size_t num_bytes) {
+inline void ShenandoahFreeSet::increase_used(size_t num_bytes) {
   shenandoah_assert_heaplocked();
   _used += num_bytes;
+  assert(_used <= _capacity, "must not use (" SIZE_FORMAT ") more than we have (" SIZE_FORMAT ") after increase by " SIZE_FORMAT,
+         _used, _capacity, num_bytes);
+}
 
-  assert(_used <= _capacity, "must not use more than we have: used: " SIZE_FORMAT
-         ", capacity: " SIZE_FORMAT ", num_bytes: " SIZE_FORMAT, _used, _capacity, num_bytes);
+inline bool ShenandoahFreeSet::peek_is_mutator_free(size_t idx) const {
+  return _mutator_free_bitmap.at(idx);
 }
 
-bool ShenandoahFreeSet::is_mutator_free(size_t idx) const {
+inline bool ShenandoahFreeSet::is_mutator_free(size_t idx) const {
   assert (idx < _max, "index is sane: " SIZE_FORMAT " < " SIZE_FORMAT " (left: " SIZE_FORMAT ", right: " SIZE_FORMAT ")",
           idx, _max, _mutator_leftmost, _mutator_rightmost);
+  assert(!_mutator_free_bitmap.at(idx) || (alloc_capacity(ShenandoahHeap::heap()->get_region(idx)) > 0),
```

Comment:
Also, for the benefit of fastdebug builds performance (= faster testing), this can be something like:

```
bool is_free = _mutator_free_bitmap.at(idx);
assert(!is_free || has_alloc_capacity(idx), "Mutator free set should contain useful regions");
return is_free;
```

Reply:
done.